src/structurizr/model/implied_relationship_strategies.py

```diff
@@ -0,0 +1,108 @@
+A also uses B (and A uses B1 and A1 uses B).  Each strategy is represented as a
+function that can be set on `Model.implied_relationship_strategy`, with the default
+being to do nothing.
+"""
+
+from itertools import product
+from typing import List
+
+from .element import Element
+from .relationship import Relationship
+from .software_system import SoftwareSystem
+
+
+def default_implied_relationship_strategy(relationship: Relationship):
+    """Don't create any implied relationships."""
+    pass
+
+
+def create_implied_relationships_unless_any_exist(relationship: Relationship):
+    """
+    Create implied relationships unless there is any existing.
+
+    This strategy creates implied relationships between all valid combinations of the
+    parent elements, unless any relationship already exists between them.
+    """
+    source = relationship.source
+    destination = relationship.destination
+    ancestor_pairs = product(_get_ancestors(source), _get_ancestors(destination))
+    for new_source, new_destination in ancestor_pairs:
+        if _implied_relationship_is_allowed(new_source, new_destination):
+            if not any(
+                r.destination is new_destination
+                for r in new_source.get_efferent_relationships()
+            ):
+                _clone_relationship(relationship, new_source, new_destination)
+
+
+def create_implied_relationships_unless_same_exists(relationship: Relationship):
+    """
+    Create implied relationships unless there is one with the same description.
+
+    This strategy creates implied relationships between all valid combinations of the
+    parent elements, unless any relationship already exists between them which has the
+    same description as the original.
+    """
+    source = relationship.source
+    destination = relationship.destination
+    ancestor_pairs = product(_get_ancestors(source), _get_ancestors(destination))
+    for new_source, new_destination in ancestor_pairs:
+        if _implied_relationship_is_allowed(new_source, new_destination):
+            if not any(
+                r.destination is new_destination
+                and r.description == relationship.description
+                for r in new_source.get_efferent_relationships()
+            ):
+                _clone_relationship(relationship, new_source, new_destination)
+
+
+def _implied_relationship_is_allowed(source: Element, destination: Element):
+    if source is destination:
+        return False
+    elif source in _get_ancestors(destination) or destination in _get_ancestors(source):
+        return False
+    return True
+
+
+def _get_ancestors(element: Element) -> List[Element]:
+    """Get the ancestors of an element, including itself."""
+    result = []
+    current = element
+    while current is not None:
+        result.append(current)
+        current = None if isinstance(current, SoftwareSystem) else current.parent
```

Comment:
Couldn't an element have more than one parent?

Reply:
No, it's a strict hierarchy (unless I'm missing something).  The potentially confusing one is that a `ContainerInstance` references the `Container`, but its parent is still considered to be the `DeploymentNode`.

It's a little irritating that `parent` isn't a property on `Element` and so `SoftwareSystem` would just have that as `None`, as that would do away with the `isinstance` test here, but that's the same in Java also.